spec/cypress/e2e/annotations_overview_spec.cy.js

```diff
@@ -0,0 +1,240 @@
+import FactoryBot from "../support/factorybot";
+import { hexToRgb } from "../support/utility";
+
+const CARD_SELECTOR = "annotation-overview-card";
+
+const LECTURE_TITLE_1 = "SageMath";
+const MEDIUM_TITLE_1 = "Intro modules";
+const LECTURE_TITLE_2 = "Lean4";
+const MEDIUM_TITLE_2 = "Intro operators";
+const MEDIUM_TITLE_3 = "Continuous functions";
+
+function createAnnotationScenario(context, userRole = "student") {
+  if (userRole === "teacher") {
+    cy.createUser("generic").as("genericUser");
+    cy.createUserAndLogin("teacher").as("user");
+  }
+  else {
+    cy.createUser("teacher").as("teacherUser");
+    cy.createUserAndLogin("generic").as("user");
+  }
+
+  // Lectures
+  cy.then(() => {
+    // a user is considered a teacher only iff they have given any lecture
+    const teacherUser = userRole === "teacher" ? context.user : context.teacherUser;
+    FactoryBot.create("lecture_with_sparse_toc", "with_title", "with_teacher_by_id",
+      { title: LECTURE_TITLE_1, teacher_id: teacherUser.id }).as("lectureSage");
+    FactoryBot.create("lecture_with_sparse_toc", "with_title", "with_teacher_by_id",
+      { title: LECTURE_TITLE_2, teacher_id: teacherUser.id }).as("lectureLean");
+  });
+
+  // Lessons
+  cy.then(() => {
+    FactoryBot.create("valid_lesson", { lecture_id: context.lectureSage.id }).as("lesson1");
+    FactoryBot.create("valid_lesson", { lecture_id: context.lectureLean.id }).as("lesson2");
+    FactoryBot.create("valid_lesson", { lecture_id: context.lectureLean.id }).as("lesson3");
+  });
+
+  // Media
+  cy.then(() => {
+    FactoryBot.create("lesson_medium", "with_video", "released",
+      "with_lesson_by_id", { lesson_id: context.lesson1.id, description: MEDIUM_TITLE_1 })
+      .as("medium1");
+    FactoryBot.create("lesson_medium", "with_video", "released",
+      "with_lesson_by_id", { lesson_id: context.lesson2.id, description: MEDIUM_TITLE_2 })
+      .as("medium2");
+    FactoryBot.create("lesson_medium", "with_video", "released",
+      "with_lesson_by_id", { lesson_id: context.lesson3.id, description: MEDIUM_TITLE_3 })
+      .as("medium3");
+  });
+
+  // Annotations
+  cy.then(() => {
+    FactoryBot.create("annotation", "with_text",
+      { medium_id: context.medium1.id, user_id: context.user.id }).as("annotation1");
+    FactoryBot.create("annotation", "with_text",
+      { medium_id: context.medium2.id, user_id: context.user.id }).as("annotation2");
+    FactoryBot.create("annotation", "with_text",
+      { medium_id: context.medium3.id, user_id: context.user.id }).as("annotation3");
+    FactoryBot.create("annotation", "with_text",
+      { medium_id: context.medium3.id, user_id: context.user.id }).as("annotation4");
+  });
+}
```

Comment:
With regards to the Cypress tests: I admit that it took way to long for me to come up with this "scenario creation" where we need a lecture, lessons and annotations. While trying out to find the right way to construct those, I thought if it wouldn't be better to create a kind of "scenario mechanism" (similar to what was available in the cypress plugin we used before).

But then the problem is that we need to verify that UI elements hold the data that it should. And not all such data is returned via FactoryBot, for example something is constructed by means of multiple fields in a model method like `def title`. If we really wanted to create "scenario Ruby files", one would have to find a way to return multiple objects at the end to the frontend and that would also be a bit of a mess with many params, e.g. in the current file: two lectures, some lessons, media and many annotations.

In the current way, whenever something is not returned by FactoryBot, we have to adjust the factory, e.g. include a `with_text` factory where we can pass in our own title. This way, we can check if the UI element holds that title, because we are in control of how it is constructed. Another possible approach could be to let FactoryBot construct the element and then later call the backend to call the `title` method on the respective elements. But that would require very specific endpoints and I'm also not too keen to create a "one fits all" cypress endpoint by letting us run arbitrary Rails commands from JS files.

It's a bit of a balancing act and with missing cypress experience I don't have the foresight to see which way will serve us "best" in the long run. With the current approach taken here, I think we are quite flexible. But there might be cases in the future, where we need more Rails functionality during the construction. Then, we could adjust the Rails helpers accordingly and if really necessary add more endpoints to cypress and hopefully those won't become rampant.

Reply:
Another thing to mention: in FactoryBot, we sometimes need to pass the full object. E.g. somewhere in the MaMpf codebase, we do something along those lines

```ruby
@course = FactoryBot...
FactoryBot.create(:lecture, course: @course)
```

or maybe
```ruby
our_teacher = FactoryBot....
FactoryBot.create(:lecture, teacher: our_teacher)
```

However, in Cypress, when we use FactoryBot to construct `our_teacher`, we get back its JSON representation. We can't use that to pass it to the `teacher` field during the lecture construction (it has to be a Ruby object). Therefore, I had to add a `with_teacher_by_id` trait to the respective factory which seems to be a bit of a "by hook or by crook" way to get things done.